packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,310 @@
+        }
+        const {exp, iat} = jwtDecode<JWTHeaders>(token.replace('Bearer ', ''))
+        const validTimeSeconds = exp - iat - 60
+        const tokenAgeSeconds = Date.now() / 1000 - iat
+        if (validTimeSeconds > tokenAgeSeconds) {
+            return false
+        }
+
+        return true
+    }
+
+    /**
+     * This function follows the public client auth flow to get
+     * access token for the user. Following the flow:
+     * 1. If we have valid access token - use it
+     * 2. If we have valid refresh token - refresh token flow
+     * 3. PKCE flow
+     *
+     * Only call this from the "ready" function, so "ready" can manage the pending state.
+     *
+     * @internal
+     */
+    private async init() {
+        if (!this.isTokenExpired(this.get('access_token'))) {
+            return this.data
+        }
+
+        const refreshToken = this.get('refresh_token_registered') || this.get('refresh_token_guest')
+
+        if (refreshToken) {
+            const res = await helpers.refreshAccessToken(this.client, {refreshToken})
+            this.handleTokenResponse(res, true)
+            return this.data
+        }
+
+        const res = await helpers.loginGuestUser(this.client, {redirectURI: this.redirectURI})
+        this.handleTokenResponse(res, true)
+
+        return this.data
+    }
+
+    private handleTokenResponse(res: ShopperLoginTypes.TokenResponse, isGuest: boolean) {
+        this.set('access_token', `Bearer ${res.access_token}`)
+        this.set('customer_id', res.customer_id)
+        this.set('enc_user_id', res.enc_user_id)
+        this.set('expires_in', `${res.expires_in}`)
+        this.set('id_token', res.id_token)
+        this.set('idp_access_token', res.idp_access_token)
+        this.set('token_type', res.token_type)
+        this.set('usid', res.usid)
+
+        const refreshTokenKey = isGuest ? 'refresh_token_guest' : 'refresh_token_registered'
+        this.set(refreshTokenKey, res.refresh_token, {
+            expires: this.REFRESH_TOKEN_EXPIRATION_DAYS
+        })
+    }
+
+    /**
+     * The ready function returns a promise, signaling whether or not
+     * the access token is avaliable. If access token is not avaliable
+     * this method will try to re-initialize.
+     *
+     * We use this method to block those commerce api calls that
+     * requires an access token.
+     */
+    async ready() {
+        if (!this.pending) {
+            this.pending = this.init()
+        }
+        return this.pending
+    }
+
+    async loginGuestUser() {
+        const redirectURI = this.redirectURI
+        const usid = this.get('usid')
+        const request = async () => {
+            const res = await helpers.loginGuestUser(this.client, {
+                redirectURI,
+                ...(usid && {usid})
+            })
+            this.handleTokenResponse(res, true)
+            return this.data
+        }
+        this.pending = request()
+        return this.pending
+    }
```

Comment:
`loginGuestUser`, `loginRegisteredUserB2C`, `logout` seem to be privileged async methods. What I mean by this is that themselves don't worry about any pending requests, then jump the queue, execute their request and set the classes `pending` request call. 

Being that this Auth class is private, you might now expect that to be a problem because you would never run 2 async methods one after another without awaiting. But because these are eventually exposed to the public via the helper actions in shopper login, the front end could do something weird. Like this:

```
const DumbComponent = () => {
    useEffect(() => {
        useShopperLogin('loginGuestUser')
        useShopperLogin('logout')
        const {data: product} = useProduct({id: '1'}) // If logout returns before the login, this will run and probably throw an error.
    }, [])

    return <div>I'm Dumb</div>
}
```

Reply:
good call.. instead of replacing `this.pending`, i guess we could queue up the requests: `this.pending.then(newAction)`.